### ISSUE TITLE
Long-lived TcpClient

### ISSUE THREAD

**Issue by hanahmily** [reactor/reactor-netty#532]
I want to use `TcpClient` to archive a reverse proxy as follows 

```java

TcpClient.create()
                                .addressSupplier(() -> InetSocketAddressUtil.createResolved(n.getT2().getHost(), n.getT2().getPort()))
                                .doOnConnected(c -> c.addHandlerLast("codec", getAppender()))
                                .handle((in, out) -> out
                                    .sendByteArray(Flux
                                        .just(n.getT1())
                                        .map(this::appendId)
                                    ))
                                .wiretap(true)
                                .connect()
                                .block()
```
`n` is a `Tuple2` object contains hostname and port of a backend server.

### Expected behavior

I tested this proxy with one backend server and hoped that `PooledConnectionProvider` sink logs like this:

```log
Created new pooled channel, now 1 active connections and 0 inactive connections
Channel acquired, now 1 active connections and 0 inactive connections
Channel acquired, now 1 active connections and 0 inactive connections
```

### Actual behavior

Actually the logs showed this

```log
Created new pooled channel, now 1 active connections and 0 inactive connections
Created new pooled channel, now 1 active connections and 1 inactive connections
Created new pooled channel, now 1 active connections and 2 inactive connections
```

After some debugging, I found that the channel was always closed by `ChannelOperations` after outbound finished writing.

### How to fix this?

Can I override TcpClient and provide a new subclass of `ChannelOperations` that doesn't close channel in `onOutboundComplete` of `ChannelOperations`. Is this a canonical practice for my case?

My codes will be like this:

```java
protected void onOutboundComplete() {
		if (log.isDebugEnabled()) {
			log.debug(format(channel(), "[{}] User Handler requesting close connection"), formatName());
		}
                // Remove markPersistent(false);
		terminate();
	}
```

### Reactor Netty version
0.8.2
### JVM version (e.g. `java -version`)
jdk1.8.0_171
### OS version (e.g. `uname -a`)
MacOS

**Comment by smaldini** [comment]
Did you try just returning Mono.never() ?

**Comment by hanahmily** [comment]
Yes, I already try this and logs
```
 Channel connected, now 1 active connections and 0 inactive connections
 Channel connected, now 2 active connections and 0 inactive connections
 Channel connected, now 3 active connections and 0 inactive connections
 Channel connected, now 4 active connections and 0 inactive connections
 Channel connected, now 5 active connections and 0 inactive connections
```

This is just like a connection leak as same as returning neverComplete()

**Comment by smaldini** [comment]
It's a short coming from us there. Let me investigate on how we can introduce that, thanks for the report !

**Comment by smaldini** [comment]
So you can as a workaround not use handle() which for now automatically close - I think we will still investigate if neverComplete should change or not tho.

So if you weren't using handle that would look like this : 

```java
Connection c = TcpClient.create()
                                .addressSupplier(() -> InetSocketAddressUtil.createResolved(n.getT2().getHost(), n.getT2().getPort()))
                                .doOnConnected(c -> c.addHandlerLast("codec", getAppender()))                            
                                .wiretap(true)
                                .connectNow();

c.outbound()
  .sendByteArray(Flux
                   .just(n.getT1())
                  .map(this::appendId))
  .then()
  .subscribe();

//later...

c.dispose();
```

**Comment by hanahmily** [comment]
I have tried this workaround, but the problem was not fixed. 

I dived into internal, then found that Connection is an object of `OperationChannels`. `c.dispose()` will close the underlying channel instead of notifying the listener(PooledConnection) a "DISCONNECTING" state message. I thought it's the reason why TcpClient didn't reuse the channel.

I also compare the implements between `ChannelOperations` and `HttpClientOperations`. In `HttpClientOperations` when inbound received last http content it invoke `terminate()`.

```java
if (msg instanceof LastHttpContent) {
			if (!started) {
				if (log.isDebugEnabled()) {
					log.debug(format(channel(), "HttpClientOperations received an incorrect end " +
							"delimiter (previously used connection?)"));
				}
				ReferenceCountUtil.release(msg);
				return;
			}
			if (log.isDebugEnabled()) {
				log.debug(format(channel(), "Received last HTTP packet"));
			}
			if (msg != LastHttpContent.EMPTY_LAST_CONTENT) {
				super.onInboundNext(ctx, msg);
			}
			//force auto read to enable more accurate close selection now inbound is done
			channel().config().setAutoRead(true);
			if (markSentBody()) {
				markPersistent(false);
			}
			terminate();
			return;
		}
```

`terminate()` will notify the listener(PooledConnection) a "DISCONNECTING" state message and does not close the underlying channle.

But `ChannelOperations` of `TcpClient` has not any ways to invoke `terminate()` without close the channel.

In conclusion, I should create a new `ChannelOperations` factory to replace the default "ops_factory". Is that true?

**Comment by haobohuiying** [comment]
@violetagg @hanahmily I am faced with the same problem in version 1.0.6， when this issue will fix

**Comment by violetagg** [comment]
@haobohuiying Which problem do you face?

Did you try instead of using `handle` to use `connection.outbound` as described here https://github.com/reactor/reactor-netty/issues/532#issuecomment-443316240?

**Comment by haobohuiying** [comment]
`ConnectionProvider provider = ConnectionProvider.builder("fixed")
                .maxConnections(20)
                .pendingAcquireTimeout(Duration.ofMillis(30000))
                .maxIdleTime(Duration.ofMillis(600))
                .maxLifeTime(Duration.ofMillis(100))
                .build();


        LoopResources loopResources = LoopResources.create("work-group", 4, false);

         Mono connectionMono =  TcpClient.create(provider)
            .runOn(loopResources)
            .host("localhost").port(6700)
            //.wiretap(true)
            .doOnChannelInit((channelObserver, channel, remoteAddress) -> {
                channel.pipeline().addBefore("reactor.right.reactiveBridge", "xml.decoder", new XmlDecoder());
                System.out.println(channel.pipeline().names());
            })
            .handle((nettyInbound, nettyOutbound) -> {
                // https://github.com/reactor/reactor-netty/issues/532
                Mono msg = Mono.never().map((v) -> {
                    return Mono.just("echo");
                });
                ByteBufFlux msgMono = ByteBufFlux.fromString(msg);

                nettyInbound.receiveObject().subscribe(obj -> {
                    if(obj instanceof XmlElementStart) {
                        XmlElementStart xmlElementStart = (XmlElementStart) obj;
                        System.out.println(xmlElementStart.name());
                    }
                    else if(obj instanceof XmlElementEnd) {
                        XmlElementEnd xmlElementEnd = (XmlElementEnd) obj;
                        System.out.println(xmlElementEnd.name());
                    }
                });

                return nettyOutbound.send(msgMono.retain());
            }).connect();

        connectionMono.subscribe();`

I use tcp client to decode xml package， bug channel always closed after sending data， after debugging， i found that the channel was closed in follow code as @smaldini descripted
`protected void onOutboundComplete() {
		if (log.isDebugEnabled()) {
			log.debug(format(channel(), "[{}] User Handler requesting close connection"), formatName());
		}
                // Remove markPersistent(false);
		terminate();
	}`

now I calls mono.never()，the channel not close，but can’t reuse 。

**Comment by violetagg** [comment]
@haobohuiying The @smaldini suggestion is that instead of using `handle`, you can just connect and then you can use `connection.inbound()` and `connection.outbound`

**Comment by Pineapplecaker** [comment]
@violetagg @hanahmily 
Hello I am also facing this problem in v1.0.6. Below is my testing code.

`
public class ReactorConnectionTest {
    private static final Logger logger = LoggerFactory.getLogger(ReactorConnectionTest.class);
    private static byte[] heartBeat = new byte[] {6,-68,0,2,0,1,69,5,79,20,67,121,0,-85,0,6,6,58,12,20,11,6};
    
    public static ConnectionProvider provider =
            ConnectionProvider.builder("fixed")
                    .maxConnections(4)
                    .pendingAcquireTimeout(Duration.ofMillis(30000))
                    .maxIdleTime(Duration.ofMillis(60))
                    .metrics(true)
                    .build();
    
    public static void main(String[] args){
        ExecutorService executor = Executors.newFixedThreadPool(5);
        Schedulers.enableMetrics();
        TestCallable task = new TestCallable();
        logger.info("Start Executing");
        executor.submit(task);
    }
    
    public static class TestCallable implements Callable<Integer> {
        @Override
        public Integer call() throws Exception {
            for(int i =0 ; i< 20; i++) {
                Connection connection =
                        TcpClient.create(provider)
                                .host("127.0.0.1")
                                .port(7777)
                                .connectNow();
                logger.info("Current Loop : {}", i);
                heartBeat[heartBeat.length - 1] = (byte) i;
                connection.outbound().sendByteArray(Flux.just(heartBeat)).then().subscribe();
                logger.info("connections: {}", connection.toString());
                logger.info("Max connections: {}", provider.maxConnections());
                logger.info("Address: {}, isPersistent: {}, channel: {}", connection.address().toString(), connection.isPersistent(), connection.channel().id() );
                
                connection.dispose();
            }
            return 1;
        }
    }
}
`

I was trying to reuse the connection from the pool and I expected the port of local address in log should remained same or limited to the amount of maxConnection. However, what I saw in the log is 

`
Current Loop : 0
connections: ChannelOperations{PooledConnection{channel=[id: 0x68e9aa45, L:/127.0.0.1:55956 - R:127.0.0.1/127.0.0.1:7777]}}
Max connections: 4
Address: 127.0.0.1/127.0.0.1:7777, isPersistent: true, channel: 68e9aa45

Current Loop : 1
connections: ChannelOperations{PooledConnection{channel=[id: 0x4ae85784, L:/127.0.0.1:55957 - R:127.0.0.1/127.0.0.1:7777]}}
Max connections: 4
Address: 127.0.0.1/127.0.0.1:7777, isPersistent: true, channel: 4ae85784

Current Loop : 2
connections: ChannelOperations{PooledConnection{channel=[id: 0xf8de7339, L:/127.0.0.1:55958 - R:127.0.0.1/127.0.0.1:7777]}}
Max connections: 4
Address: 127.0.0.1/127.0.0.1:7777, isPersistent: true, channel: f8de7339

Current Loop : 3
connections: ChannelOperations{PooledConnection{channel=[id: 0x4b498782, L:/127.0.0.1:55959 - R:127.0.0.1/127.0.0.1:7777]}}
Max connections: 4
Address: 127.0.0.1/127.0.0.1:7777, isPersistent: true, channel: 4b498782

Current Loop : 4
connections: ChannelOperations{PooledConnection{channel=[id: 0x8448061e, L:/127.0.0.1:55960 - R:127.0.0.1/127.0.0.1:7777]}}
Max connections: 4
Address: 127.0.0.1/127.0.0.1:7777, isPersistent: true, channel: 8448061e

Current Loop : 5
connections: ChannelOperations{PooledConnection{channel=[id: 0x9a9b71af, L:/127.0.0.1:55961 - R:127.0.0.1/127.0.0.1:7777]}}
Max connections: 4
Address: 127.0.0.1/127.0.0.1:7777, isPersistent: true, channel: 9a9b71af

Current Loop : 6
connections: ChannelOperations{PooledConnection{channel=[id: 0xaafa761b, L:/127.0.0.1:55962 - R:127.0.0.1/127.0.0.1:7777]}}
Max connections: 4
Address: 127.0.0.1/127.0.0.1:7777, isPersistent: true, channel: aafa761b

Current Loop : 7
connections: ChannelOperations{PooledConnection{channel=[id: 0x6e3d6e45, L:/127.0.0.1:55963 - R:127.0.0.1/127.0.0.1:7777]}}
Max connections: 4
Address: 127.0.0.1/127.0.0.1:7777, isPersistent: true, channel: 6e3d6e45

Current Loop : 8
connections: ChannelOperations{PooledConnection{channel=[id: 0xa389aa3b, L:/127.0.0.1:55964 - R:127.0.0.1/127.0.0.1:7777]}}
Max connections: 4
Address: 127.0.0.1/127.0.0.1:7777, isPersistent: true, channel: a389aa3b

Current Loop : 9
connections: ChannelOperations{PooledConnection{channel=[id: 0x69f11144, L:/127.0.0.1:55965 - R:127.0.0.1/127.0.0.1:7777]}}
Max connections: 4
`

The port keep changing every time looks like the connection pool doesn't work. It open and close a connection every time without reusing it. 

Do I misunderstand something or that's the right behavior of the ConnectionProvider ?

**Comment by violetagg** [comment]
@Pineapplecaker `connection.dispose()` closes the connection does not return it to the pool.

**Comment by violetagg** [comment]
The current implementation does now allow you to return the connection to the pool as Reactor Netty does not know in what state the connection is.

**Comment by Pineapplecaker** [comment]
So, then currently I have to use other library to achieve the return connection to pool. Will it become a new feature in the future ? Or it is the essence that Reactor Netty is not able to know what the state of the connection is.

**Comment by violetagg** [comment]
> So, then currently I have to use other library to achieve the return connection to pool. 

yes

> Will it become a new feature in the future ? Or it is the essence that Reactor Netty is not able to know what the state of the connection is.

basically the applications/frameworks know what kind of protocol for communication is used and whether and when it is save to return the connection to the pool.

**Comment by violetagg** [comment]
I added in the reference documentation an example with `Connection#outbound` and `Connection#inbound` as an alternative of I/O handler.
I'm closing this issue as for the time being there are no plans for allowing to release back to the pool a tcp connection.

**Comment by gbhatia81** [comment]
@violetagg I think it would help many people if we clearly document that for TCP connection, connections are not released back to pool. I understand the reason and it makes sense as well but since documentation was not clearly mentioning same, i had to spend many hours in code to figure out that that is the case. Documentation on contrary has a connection pool section within TCP client and this creates more confusion.

**Comment by violetagg** [comment]
@gbhatia81 Would you like to make a PR for the documentation?

**Comment by violetagg** [comment]
I updated the documentation. See #1999

**Comment by a-baki** [comment]
> I added in the reference documentation an example with `Connection#outbound` and `Connection#inbound` as an alternative of I/O handler. I'm closing this issue as for the time being there are no plans for allowing to release back to the pool a tcp connection.

Yes, this way it retains the connection but I ve noticed 2 flaws pertaining to this usage,
1 there is no point in using a pool if only one connection/channel can be utilized this way, pooling works but actually works for nothing. 
2 I found out that also channels lifetime is bound to inbound read time out. it acts like pool idle timeout if you want the connection to be alive, after read timeout occurs channel is also disposed. That is also a problem,  you have to give a long read timeout value in order to keep channel alive. that can cause lo

I tried to implement a simple client and server, where server echoes the input back to client, but I cant find a way to implement this simple conversation using TcpClient with connection pooling. Seems that you have to override default behavior with alternate implementation as @hanahmily mentioned. Also I agree that it should be achievable with configuration.
